### ISSUE TITLE
Roadmap and future plans

### ISSUE THREAD

**Issue by abastardi** [mattkrick/meatier#41]
What are your plans for this project? It seems like it has a lot of potential.

Are you interested in turning it into something more formal (maybe like [MEAN.JS](http://meanjs.org/)) and building a community around it?

Any roadmap?

**Comment by mattkrick** [comment]
I think the days of stacks like MEAN are over. How many people use "MEAN" with Koa or React or Posgresql? The JS world is moving waaaaay too fast to coin something that has the sticking power that LAMP had.

If a community forms around it, that'd be amazing. Currently, there is not a single example of a production-ready realtime app, that's the niche I want to fill. 

What I don't want to do is compete with Meteor. Their niche is MVPs and first-time programmers. I think folks just stick with Meteor until they hit a point in their project/career where they need more flexibility. I also don't want any proprietary junk, like Meteor.methods. That locks people in, which isn't cool. I've only written 2 packages to make this work: `redux-optimistic-ui` and `redux-socket-cluster`. Neither are Meatier-specific, which keeps the ecosystem modular.

**Comment by abastardi** [comment]
> I think the days of stacks like MEAN are over. How many people use "MEAN" with Koa or React or Posgresql? The JS world is moving waaaaay too fast to coin something that has the sticking power that LAMP had.

I agree. Note, I wasn't suggesting committing to a fixed stack like MEAN, but simply the idea of developing something a little more formal than just a boilerplate/example app. In other words, extract out the common, generic core that can be used like a framework and provide some documentation and guidance. Also, maybe at some point a more formal presentation, like a separate website.

By the way, end-user documentation would be great, but just as helpful (maybe moreso in the early stages) would be a walk-through of the underlying code and how everything works -- something that would help potential contributors better comprehend the system and make it easier to jump in.

> If a community forms around it, that'd be amazing. Currently, there is not a single example of a production-ready realtime app, that's the niche I want to fill.

+1

At some point, you might consider a new name. "Meatier" is a cute pun, but if this develops into something more serious, it should be thought of as something that stands on its own rather than a joke alternative to the real Meteor.

**Comment by mattkrick** [comment]
Haha yeah, if it got serious I'd change the name, but at ~300 stars, it's a long ways from anything special.

When you say "framework" what would you extract? IMO "framework" doesn't exist. Everything can be swapped out or modified, whether it's the server, the database, or the auth boilerplate. Therefore an underlying framework doesn't exist. If something *could* be swapped out, it'd be its own NPM package. That's the beauty (and the pain); everything is modular. 

In Meteor, the pubsub, the auth, the builder, tracker... it's all part of the framework. What if you want to edit an auth template? *You have to write a blaze template that overwrites a default one & send them both to the client* Bloat is gross. What if you want to do something fun like automatically log someone in from the signup page if their username & password match an existing user? You ***can't***.

**Comment by abastardi** [comment]
> When you say "framework" what would you extract? IMO "framework" doesn't exist. Everything can be swapped out or modified, whether it's the server, the database, or the auth boilerplate.

Sure, it's not a framework in the same sense that Meteor is, but you're written a lot of generalized code that can be used beyond the Kanban example app. So, for starters, I would say separate out all of the Kanban specific code, and consider what's left the "framework" (or rather, the "glue" that brings all the NPM pieces together into a coherent whole).

Specifically, maybe move the Kanban specific code into its own folder. This would include Kanban specific files (such as /server/publish/allNotes.js) as well Kanban specific code within more general files (such as some of the [socket setup code](https://github.com/mattkrick/meatier/blob/master/src/server/worker.js#L66-L71)  in /server/worker.js). Leave the auth code as part of the "framework".

As an additional step, maybe pull out some of the React components into an emtpy front-end scaffold as a starter template (i.e., minus the Kanban stuff).

Finally, consider adding some helpers to reduce boilerplate for common patterns, such as the model and controller code for adding, updating, and deleting documents.

So, basically, just a cleaner separation between (1) the core reusable parts, (2) a basic front-end scaffold, and (3) the Kanban specific code. I think this would make it easier to understand, to use, and to make contributions.

**Comment by mattkrick** [comment]
that's good, in my real project based on meatier i break out landing page, auth, and portal (think kanban) into 3 separate modules, each with their own components, containers, redux, etc. probably worthwhile to port over that same folder structure into this. 

What are your thoughts for helpers for Add/update/delete? I've attempted that before, but always failed because they're so dependent on the relational schema. PR welcomed if you come up with a cleaner solution!

**Comment by abastardi** [comment]
> What are your thoughts for helpers for Add/update/delete? I've attempted that before, but always failed because they're so dependent on the relational schema. PR welcomed if you come up with a cleaner solution!

I was just noticing that /models/notes.js and /models/lanes.js are basically identical, and the same for /controllers/notes.js and /controllers/lanes.js. And none of those files include any schema-specific code. So I was thinking at least for the simple case of adding, updating, deleting a single document in a table, all of that boilerplate could be abstracted into a few simple helpers, with the relevant Thinky model or Joi schema passed in as an argument.

**Comment by mattkrick** [comment]
Nah, it'd be hard to squash em as is since deleting a lanes deletes all notes inside. Give a PR a shot & I think you'll see what I mean :wink:

**Comment by abastardi** [comment]
Sorry, maybe I'm missing something, but was just thinking of something like this:

    export async function deleteDocDB(model, id) {
      let docToDelete;
      try {
        docToDelete = await model.get(id);
      } catch (e) {
        throw e;
      }
      try {
        await docToDelete.delete();
      } catch (e) {
        throw e;
      }
    }

    export async function deleteDoc(payload, model, callback) {
      const {id} = payload;
      this.docQueue.add(id);
      try {
        await deleteDocDB(model, id);
      } catch (e) {
        this.docQueue.delete(id);
        return callback(null, {_error: e.message})
      }
      callback();
    }

Then in the app, replace all of that code with:

    export async function deleteLane(payload, callback) {
      deleteDoc(payload, Lane, callback);
    }

Where `Lane` is define as before.

I'm sure this can be improved, but the idea is just to remove all the repetitive boilerplate.

**Comment by mattkrick** [comment]
can that handle deleting all notes associated with lanes?
what if notes & lanes has a M:N relationship?

**Comment by abastardi** [comment]
> but then how do you propagate the correct redux action type back to the client?

My example runs all the same code that the app currently runs (just moving some of the boilerplate into a couple of abstracted functions), so nothing else should change. Presumably, the post-database-operation redux action will continue to be handled via the [callback](https://github.com/mattkrick/meatier/blob/77e6715423e3e40ac9f50b7bb83ecdbcc2a12395/src/universal/redux/middleware/optimisticMiddleware.js#L22-L30) in optimisticMiddleware.js.

**Comment by mattkrick** [comment]
Got rid of all the things talked about here when I switched to graphQL, still a lot of boilerplate, but having an entire schema laid out on the server is really nice and will pay dividends for larger apps (although it looks pretty silly using graphQL for something tiny like this)
